[SSWG-0020] Distributed actors (cluster) (#72)
swift-server/sswg: proposals/0020-distributed-actor-cluster.md

This package contains a feature rich implementation of an actor system cluster, and general framework for building distributed systems in Swift. It integrates with various observability libraries the SSWG endorses and is based on top of swift-nio for its networking stack.

|  |  |
|--|--|
| **Package Name** | `swift-distributed-actors` |
| **Module Name** | `DistributedCluster` |
| **Proposed Maturity Level** | [Sandbox](https://github.com/swift-server/sswg/blob/main/process/incubation.md#process-diagram) |
| **License** | [Apache 2.0](https://github.com/swift-server/swift-backtrace/blob/master/LICENSE.txt) |
| **Dependencies** | [swift-nio](https://github.com/apple/swift-nio), [swift-log](https://github.com/apple/swift-log), [swift-metrics](https://github.com/apple/swift-metrics), [swift-service-discovery](https://github.com/apple/swift-service-discovery), [swift-cluster-membership](https://github.com/apple/swift-cluster-membership) |

## Introduction

Distributed actors are an extension of the "local only" actor model offered by Swift with its `actor` keyword.

> Distributed actors used by `distributed actor` keywords (and importing the `Distributed` module, provided by Swift), and enable the declaring of `distributed func` methods inside such actor. Such methods may then be invoked remotely, from other peers in a distributed actor system.

Reviewer: Wording seems slightly off
```suggestion
Distributed actors used the `distributed actor` keywords (after importing the `Distributed` module, provided by Swift), and enable the declaring of `distributed func` methods inside such actor. Such methods may then be invoked remotely, from other peers in a distributed actor system.
```

The distributed actor *language feature* does not include any specific *runtime*, and only defines the language and semantic rules surrounding distributed actors. This library provides a feature-rich clustering server-side focused implementation of such runtime (i.e. a `DistributedActorSystem` implementation) for distributed actors.

Clusters are a fundamental building block of server-side distributed systems, such as lobby systems for game backends, IoT service backends, orchestration systems and general control plane systems of distributed databases or other distributed systems.

## Motivation

> Distributed actors are introduced as nominal types in Swift 5.7. Similarly to actors, they can be declared using the `distributed actor` pair of keywords. By themselves, they cannot really do anything – all the distributed actions such actor performs are actually handled an `ActorSystem` associated with given actor type. 

Reviewer: ```suggestion
Distributed actors are introduced as nominal types in Swift 5.7. Similar to actors, they can be declared using the `distributed actor` pair of keywords. By themselves, they cannot really do anything – all the distributed actions such actor performs are actually handled by an `ActorSystem` associated with given actor type. 
```

```swift
typealias DefaultDistributedActorSystem = ClusterSystem
```

Which avoids having to repeat the `ActorSystem` typealias in every distributed actor in a module.

The package being proposed here provides the implementation of the `ClusterSystem`.

## Proposed solution

The `DistributedCluster` package includes the `ClusterSystem` type which is the central piece of the library. Once created, it binds to a host/port pair and begins listening for incoming connections:

```swift
@main
struct Main {
    static func main() async throws {
        let system = await ClusterSystem("FirstSystem") { settings in
            settings.endpoint.host = "127.0.0.1"
            settings.endpoint.port = 7337
        }
        
        try await system.terminated
    }
}
```

Clusters are formed by “joining” other nodes. So the above created node has bound on port 7337, and if we had another node running on the same host but on port 8228 we could join in programmatically like this:

```swift
let other = Cluster.Endpoint(host: "127.0.0.1", port: 8228)
system.cluster.join(endpoint: other)
```

> The `join(endpoint:)` method returns immediately, however it is possible to `try await joined(endpoint:within:)` to suspend until the other node has been joined (or we failed doing so within the timeout).

Reviewer: ```suggestion
The `join(endpoint:)` method returns immediately, however it is possible to use `try await joined(endpoint:within:)` to suspend until the other node has been joined (or we failed doing so within the timeout).
```

```swift
import ServiceDiscovery
import K8sServiceDiscovery // See: [tuplestream/swift-k8s-service-discovery](https://github.com/tuplestream/swift-k8s-service-discovery)
import DistributedCluster

ClusterSystem("DiscoverNodes") { settings in
    let discovery = K8sServiceDiscovery() 
    let target = K8sObject(labelSelector: ["name": "actor-cluster"], namespace: "actor-cluster")
    
    settings.discovery = ServiceDiscoverySettings(discovery, service: target)
}
```

Once a cluster has been formed, distributed actors use the [Receptionist](https://apple.github.io/swift-distributed-actors/1.0.0-beta.2/documentation/distributedactors/receptionist) to discover check-in and discover distributed actors from other nodes. For example, we can write a distributed worker actor like this:

```swift
distributed actor Worker {
    typealias ActorSystem = ClusterSystem
    
    init(actorSystem: ActorSystem) async {
        self.actorSystem = actorSystem

>         await system.receptionist.checkIn(worker, with: .workers) 

Reviewer: where is `system` from here - is it a singleton?

Author: typo, thanks!

It is supposed to be `actorSystem`; my muscle memory won here hehe... I often call it system in the lib

```suggestion
        await actorSystem.receptionist.checkIn(worker, with: .workers) 
```

    }
    
    distributed func work() -> String { "Completed some work!" }
}

extension DistributedReception.Key {
    static var workers: DistributedReception.Key<Worker> {
        "workers"
    }
}
```

> And have all workers that are initialized on any node that has joined the cluster by using the [receptionist listing](https://apple.github.io/swift-distributed-actors/1.0.0-beta.2/documentation/distributedactors/receptionist#Receptionist-Listings):

Reviewer: This sentence needs refining

Author: Thanks will do

```suggestion
Such distributed actors are then created using normal initialization on nodes within the cluster. Each such node can then use the [receptionist listing](https://apple.github.io/swift-distributed-actors/1.0.0-beta.3/documentation/distributedactors/receptionist#Receptionist-Listings) API to obtain all workers from all other nodes, like this:
```
